bitcoin/src/lib.rs

```diff
@@ -242,3 +242,25 @@ mod encode_impls {
     impl_encodable_for_u32_wrapper!(BlockHeight);
     impl_encodable_for_u32_wrapper!(BlockInterval);
 }
+
+/// A conversion trait for unsigned integer types smaller than or equal to 64-bits.
+///
+/// This trait exists because [`usize`] doesn't implement `Into<u64>`, however because we only
+/// support 32 and 64 bit architectures we can infallibly do the conversion.
+pub trait ToU64 {
```

Comment:
I really don't like dumping random shit into crate root. This should either go to `internals` or shouldn't be public. `internals` would also take care of compile errors in the previous commit - internals would be compiled first and fail. Though maybe it's too restrictive since `units` could actually support 16-bit.

Reply:
If `units` wants to support 16-bit processors it can just avoid using this macro. I think we should default to blocking 16-bit systems except in crates where we take active steps to support them.

Which is an argument for pulling this trait into internals and then making sure any casts to `usize` have a good justification for not using it, and then in `units` if we want we can use an alternate trait.

Reply:
I think the problem is this trait is doing two things:
1. Provide a function so we don't have to cast (this is a totally internal thing)
2. Provide a trait for the `consensus::encode` module to use (this is a public thing)

In #2931 we want to do something like:

```rust
fn emit_varint(&mut self, v: impl ToU64) -> Result<usize, io::Error>;
```

It also turns out that we only do suspicious casting (at the moment) in `bitcoin` (note patch 2 in this PR only touches `bitcoin/`).

Thoughts please?

Reply:
> I think we should default to blocking 16-bit systems except in crates where we take active steps to support them.

Yes, but then we should make the trait optional and only available with `only_32_and_64_bit_systems` feature flag which all our crates will enable. And we should make sure `units` API is ready for 16-bits by not using the trait in the API.

@tcharding if we seal the trait and decide to only implement it for those integers it's already implemented for then the problems of the trait being in a stable crate are very minimal. The only real problem I can think of is `crate_a` defining `fn foo(x: impl ToU64) { crate_b::bar(x) }` where `bar`, then if one of the crates updates this breaks. We could advise early conversions (which are good for performance anyway) to guard against this: `foo(x: impl ToU64) { crate_b::bar(x.to_u64()) }`. Maybe this is not a long-term solution though and we need a stable crate.

Reply:
Perhaps this: https://github.com/rust-bitcoin/rust-bitcoin/discussions/2938

Reply:
I don't think we should make every single user and maintainer need to think about an extra feature just so that some crates in the ecosystem are able to work on 16-bit systems.

I wouldn't mind having a *cfg* flag to *enable* 16-bit compilation, and having most crates fail compilation if it's enabled. Though that does mean that we need to add a few lines of code to every crate in order to prevent compilation in 16-bit mode.

Reply:
@apoelstra I meant it to be in `internals`, so downstream users don't need to worry about it. `units`, `address` etc would turn it off, so they can still use `internals` without breaking compilation other crates would enable it to trigger the error message.

Actually inverting it would cause more work for downstream users. With the feature as I described the users will get usable `units` and the code compiles because they don't depend on other crates with no additional work.